File: Integraupt-Backend/BackendInicidencia/src/main/java/com/inicidencia/Controller/IncidenciaController.java
package com.inicidencia.controller;

import com.inicidencia.dto.DisponibilidadIncidenciaResponse;
import com.inicidencia.dto.IncidenciaGestionResponse;
import com.inicidencia.dto.IncidenciaRequest;
import com.inicidencia.dto.IncidenciaResponse;
import com.inicidencia.service.IncidenciaService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import com.inicidencia.dto.ReservaIncidenciaResumen;
import org.springframework.web.bind.annotation.CrossOrigin;

import java.util.List;

@RestController
@RequestMapping("/api/incidencias")
@CrossOrigin(origins = "*")
public class IncidenciaController {

    private final IncidenciaService incidenciaService;

    public IncidenciaController(IncidenciaService incidenciaService) {
        this.incidenciaService = incidenciaService;
    }

    @PostMapping
    public ResponseEntity<IncidenciaResponse> registrarIncidencia(@Valid @RequestBody IncidenciaRequest request) {
        IncidenciaResponse incidencia = incidenciaService.registrarIncidencia(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(incidencia);
    }

    @GetMapping
    public List<IncidenciaGestionResponse> listarParaGestion(
            @RequestParam(value = "rol", required = false) String rol,
            @RequestParam(value = "facultadId", required = false) Long facultadId,
            @RequestParam(value = "escuelaId", required = false) Long escuelaId,
            @RequestParam(value = "escuelaContextoId", required = false) Long escuelaContextoId,
            @RequestParam(value = "espacioId", required = false) Integer espacioId,
            @RequestParam(value = "search", required = false) String search
    ) {
        return incidenciaService.listarIncidenciasParaGestion(
                rol,
                facultadId,
                escuelaId,
                escuelaContextoId,
                espacioId,
                search
        );
    }

    @GetMapping("/reserva/{reservaId}")
    public List<IncidenciaResponse> listarPorReserva(@PathVariable Integer reservaId) {
        return incidenciaService.listarPorReserva(reservaId);
    }

    @GetMapping("/reserva/{reservaId}/disponibilidad")
    public DisponibilidadIncidenciaResponse verificarDisponibilidad(@PathVariable Integer reservaId) {
        return incidenciaService.verificarDisponibilidad(reservaId);
    }
    @GetMapping("/usuario/{usuarioId}/reservas")
    public List<ReservaIncidenciaResumen> listarReservasPorUsuario(@PathVariable Integer usuarioId) {
        return incidenciaService.listarReservasParaUsuario(usuarioId);
    }

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<String> manejarNoEncontrado(IllegalArgumentException ex) {
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(ex.getMessage());
    }

    @ExceptionHandler(IllegalStateException.class)
    public ResponseEntity<String> manejarEstadoInvalido(IllegalStateException ex) {
        return ResponseEntity.status(HttpStatus.CONFLICT).body(ex.getMessage());
    }
}

File: Integraupt-Backend/backend-incidencia/app/Http/Controllers/IncidenciaController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use App\Http\Requests\IncidenciaRequest;
use App\Services\IncidenciaService;
use Illuminate\Database\Eloquent\ModelNotFoundException;

class IncidenciaController extends Controller
{
    private $service;

    public function __construct(IncidenciaService $service)
    {
        $this->service = $service;
    }

    public function registrarIncidencia(IncidenciaRequest $request)
    {
        try {
            $incidencia = $this->service->registrarIncidencia($request->validated());
            return response()->json($incidencia, 201);

        } catch (\InvalidArgumentException | ModelNotFoundException $e) {
            $message = $e instanceof ModelNotFoundException ? "No se encontró la reserva indicada." : $e->getMessage();
            return response()->json($message, 404);

        } catch (\LogicException $e) {
            return response()->json($e->getMessage(), 409);
            
        } catch (\Exception $e) {
            return response()->json($e->getMessage(), 500);
        }
    }

    public function listarParaGestion(Request $request)
    {
        return response()->json(
            $this->service->listarIncidenciasParaGestion(
                $request->rol,
                $request->facultadId,
                $request->escuelaId,
                $request->escuelaContextoId,
                $request->espacioId,
                $request->search
            )
        );
    }

    public function listarPorReserva($reservaId)
    {
        return response()->json(
            $this->service->listarPorReserva($reservaId)
        );
    }

    public function verificarDisponibilidad($reservaId)
    {
        return response()->json(
            $this->service->verificarDisponibilidad($reservaId)
        );
    }

    public function listarReservasPorUsuario($usuarioId)
    {
        return response()->json(
            $this->service->listarReservasParaUsuario($usuarioId)
        );
    }
}